runtime/signals.c

```diff
@@ -289,50 +300,78 @@ void caml_request_minor_gc (void)
+  if (Caml_state->requested_major_slice || Caml_state->requested_minor_gc) {
+    CAML_INSTR_INT ("force_minor/caml_process_pending_gc@", 1);
     caml_gc_dispatch();
   }
-  CAMLreturn (extra_root);
 }
 
-CAMLexport value caml_check_urgent_gc (value extra_root)
+value caml_process_pending_gc_with_root(value extra_root)
 {
   CAMLparam1 (extra_root);
+  caml_process_pending_gc();
+  CAMLreturn (extra_root);
+}
+
+CAMLexport value caml_process_pending_events_exn (void)
+{
   caml_something_to_do = 0;
 #ifdef NATIVE_CODE
   caml_update_young_limit();
 #endif
-  if (Caml_state->requested_major_slice || Caml_state->requested_minor_gc){
-    CAML_INSTR_INT ("force_minor/check_urgent_gc@", 1);
-    caml_gc_dispatch();
-  }
-  caml_memprof_handle_postponed();
-  caml_final_do_calls();
-  caml_process_pending_signals();
-  CAMLreturn (extra_root);
+  caml_process_pending_gc();
+
+  value res;
+
+  // Call memprof callbacks
+  res = caml_memprof_handle_postponed_exn();
+  if (Is_exception_result(res)) goto exception;
+
+  // Call finalisers
+  res = caml_final_do_calls_exn();
+  if (Is_exception_result(res)) goto exception;
+
+  // Call exception handlers
+  res = caml_process_pending_signals_exn();
+  if (Is_exception_result(res)) goto exception;
+
+  return Val_unit;
+
+exception:
+  /* If an exception is raised during an asynchronous callback, then
+     it might be the case that we did not run all the callbacks we
+     needed. Therefore, we set [caml_something_to_do] again in order
+     to force reexamination of callbacks. */
+  caml_set_urgent_to_do();
+  return res;
+}
+
+CAMLexport void caml_process_pending_events(void)
+{
+  value res = caml_process_pending_events_exn();
+  if (Is_exception_result(res)) caml_raise(Extract_exception(res));
 }
 
+CAMLexport value caml_process_pending_events_with_root_exn(value extra_root)
+{
+  CAMLparam1 (extra_root);
+  value res = caml_process_pending_events_exn();
+  CAMLreturn (Is_exception_result(res) ? res : extra_root);
+}
 
-/* If an exception is raised during an asynchronous callback (i.e.,
-   from an OCaml function called in `caml_check_urgent_gc`), then it
-   might be the case that we did not run all the callbacks we needed
-   to run even though [caml_something_to_do] has been reset to 0 at
-   the begining of [caml_check_urgent_gc]. Therefore, we set
-   [caml_something_to_do] in order to force reexamination of
-   callbacks.
+CAMLexport value caml_process_pending_events_with_root(value extra_root)
+{
+  CAMLparam1 (extra_root);
+  caml_process_pending_events();
+  CAMLreturn (extra_root);
+}
 
-   Apart from a reasonable performance penalty (an extra call to
-   `caml_check_urgent_gc`), it is OK to call this function where
-   `caml_raise` would have been more appropriate (i.e., not called
-   from `caml_check_urgent_gc`). */
-void caml_raise_in_async_callback (value exc)
+// backwards-compatibility
+CAMLexport value caml_check_urgent_gc(value extra_root)
```

Comment:
The use of macros in the codebase leaves me cold. Maybe using the C99 inline keyword would be better? Is there any guidance on the use of macros in the OCaml codebase?

Reply:
I don't think there is any general guidance. I would say that using macros for defining aliases is fine, and this is what `compatibility.h` has done for years for many symbols.

That said, I do not care too strongly, and this may even become impossible if you take into account my request of merging the various functions.